test/boost/gcp_object_storage_test.cc

```diff
@@ -0,0 +1,456 @@
+    _Never_ give any real priviledges to these accounts, as we are obviously exposing them here.
+
+    User1 is assumed to have permissions to read/write the bucket
+    User2 is assumed to only have permissions to read the bucket, but permission to 
+    impersonate User1.
+
+    Note: fake gcp storage does not have any credentials or permissions, so
+    for testing with such, leave them unset to skip those tests.
+
+    This test is parameterized with env vars:
+    * ENABLE_GCP_STORAGE_TEST - set to non-zero (1/true) to run
+    * GCP_STORAGE_ENDPOINT - set to endpoint host. default is https://storage.googleapis.com
+    * GCP_STORAGE_PROJECT - project in which to create bucket (if not specified)
+    * GCP_STORAGE_USER_1_CREDENTIALS - set to credentials file for user1
+    * GCP_STORAGE_USER_2_CREDENTIALS - set to credentials file for user2
+    * GCP_STORAGE_BUCKET - set to test bucket
+*/
+
+struct gcp_test_env {
+    std::string endpoint;
+    std::string project;
+    std::string bucket; 
+    std::string user_1_creds;
+    std::string user_2_creds;
+    // for test
+    std::vector<std::string> objects_to_delete;
+};
+
+static std::string get_var_or_default(const char* var, std::string_view def) {
+    const char* val = std::getenv(var);
+    if (val == nullptr) {
+        return std::string(def);
+    }
+    return val;
+}
+
+using test_func = std::function<future<>(const tmpdir&, storage::client&, gcp_test_env&)>;
+
+static future<std::optional<google_credentials>> credentials(const std::string& source) {
+    try {
+        if (!source.empty()) {
+            BOOST_TEST_MESSAGE(fmt::format("Loading credentials from {}", source));
+            co_return co_await google_credentials::from_file(source);
+        }
+        BOOST_TEST_MESSAGE("Loading default credentials");
+        co_return co_await google_credentials::get_default_credentials();
+    } catch (...) {    
+        BOOST_TEST_MESSAGE(fmt::format("Warning: could not load {} credentials: {}", source, std::current_exception()));
+    }
+    co_return std::nullopt; // empty, useless. will work on fake server 
+}
+
+namespace fs = std::filesystem;
+namespace tp = tests::proc;
+
+static future<std::tuple<tp::process_fixture, int>> start_fake_gcs_server(const tmpdir& tmp) {
+    std::vector<std::string> params({
+        "", "-filesystem-root", (tmp.path() / "objects").string()
+    });
+
+    size_t port_index = 0;
+
+    auto exec = tp::find_file_in_path("fake-gcs-server");
+    if (exec.empty()) {
+        exec = tp::find_file_in_path("docker");
+        if (exec.empty()) {
+            exec = tp::find_file_in_path("podman");
+        }
+        if (exec.empty()) {
+            throw std::runtime_error("Could not find fake-gcs-server or docker.");
+        }
+
+        // publish port ephemeral, allows parallel instances
+        params = {
+            "",
+            "run", "--rm", "-p", "--", // set below
+            "docker.io/fsouza/fake-gcs-server"
```

Comment:
Please use a tag here (e.g. :1.52.3) so we aren't exposed to random regressions.

Reply:
Done.